tensorflow/core/kernels/roll_op_gpu.cu.cc

```diff
@@ -0,0 +1,103 @@
+/* Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#if GOOGLE_CUDA
+
+#define EIGEN_USE_GPU
+
+#include "tensorflow/core/kernels/roll_op.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/core/util/cuda_kernel_helper.h"
+
+namespace tensorflow {
+
+typedef Eigen::GpuDevice GPUDevice;
+
+namespace {
+
+template <typename T>
+__global__ void RollKernel(
+    const int32 nthreads, const int32 num_dims,
+    const T* input, T* output,
+    const int32* dim_size, const int32* threshold, const int64* dim_range) {
+  CUDA_1D_KERNEL_LOOP(out_idx, nthreads) {
+    int64 offset = 0;
+    for (int i = 0; i < num_dims; i++) {
+        const int64 stride = dim_range[i] / dim_size[i];
+        const int shift = dim_size[i] - threshold[i];
+        const int indx = (out_idx / stride) % dim_size[i];
+        const int shifted_indx = (indx + shift) % dim_size[i];
+        offset += (shifted_indx - indx) * stride;
+    }
+    output[out_idx + offset] = input[out_idx];
+  }
+}
+}  // namespace
+
+namespace functor {
+
+
+template <typename T>
+struct Roll<GPUDevice, T> {
+  void operator()(const OpKernelContext *context,
+                        const int64 num_elements,
+                        const int num_dims,
+                        const gtl::ArraySlice<int32> dim_size,
+                        const T *input, T *output,
+                        const gtl::ArraySlice<int32> threshold,
+                        const gtl::ArraySlice<int64> dim_range,
+                        const int64 isd) {
+    const GPUDevice& d = context->eigen_device<GPUDevice>();
+
+    auto dim_bytes = sizeof(int32) * dim_size.size();
+    auto dim_buf = d.allocate(dim_bytes);
+
+    auto thres_bytes = sizeof(int32) * threshold.size();
+    auto thres_buf = d.allocate(thres_bytes);
+
+    auto range_bytes = sizeof(int64) * dim_range.size();
+    auto range_buf = d.allocate(range_bytes);
+
+    d.memcpyHostToDevice(dim_buf, dim_size.data(), dim_bytes);
+    d.memcpyHostToDevice(thres_buf, threshold.data(), thres_bytes);
+    d.memcpyHostToDevice(range_buf, dim_range.data(), range_bytes);
```

Comment:
I'm pretty sure this would be faster if you allocated and copied them in one go. You would need to merge the arrays first and make sure the underlying storage (preferably page-locked) remains allocated.

Actually, I think the lifetime is not guaranteed in the current code either. Feel free to ignore the optimization, but please check correctness.

Reply:
Here, I follow the example of transpose_functor_gpu.cu.cc for memory allocation
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/transpose_functor_gpu.cu.cc

I run some unit tests, which are correct.

Reply:
Indeed, allocate+memcpy+deallocate is a valid and correct pattern. The deallocation is inserted into the stream and not carried out immediately.